src/Features/Core/Portable/Completion/Providers/ImportCompletionProvider/AbstractExtensionMethodImportCompletionProvider.cs

```diff
@@ -0,0 +1,277 @@
+﻿// Copyright (c) Microsoft.  All Rights Reserved.  Licensed under the Apache License, Version 2.0.  See License.txt in the project root for license information.
+
+using System;
+using System.Collections.Generic;
+using System.Collections.Immutable;
+using System.Diagnostics;
+using System.Linq;
+using System.Threading;
+using System.Threading.Tasks;
+using Microsoft.CodeAnalysis.Completion.Log;
+using Microsoft.CodeAnalysis.FindSymbols;
+using Microsoft.CodeAnalysis.PooledObjects;
+using Microsoft.CodeAnalysis.Shared.Extensions;
+using Microsoft.CodeAnalysis.Shared.Extensions.ContextQuery;
+using Roslyn.Utilities;
+
+namespace Microsoft.CodeAnalysis.Completion.Providers
+{
+    internal class ExtensionMethodImportCompletionService
+    {
+        // return a multi-dictionary of mappings "FQN of containing type" => "extension method names"
+        public async Task<MultiDictionary<string, string>> GetPossibleExtensionMethodMatchesAsync(
+            Project currentProject,
+            ISet<string> targetTypeNames,
+            CancellationToken cancellationToken)
+        {
+            var solution = currentProject.Solution;
+            var graph = currentProject.Solution.GetProjectDependencyGraph();
+            var dependencies = graph.GetProjectsThatThisProjectTransitivelyDependsOn(currentProject.Id);
+            var relevantProjects = dependencies.Select(solution.GetProject)
+                                                 .Where(p => p.SupportsCompilation)
+                                                 .Concat(currentProject);
+
+            var results = new MultiDictionary<string, string>();
+
+            // Find matching extension methods from source.
+            foreach (var project in relevantProjects)
+            {
+                foreach (var document in project.Documents)
+                {
+                    var info = await document.GetSyntaxTreeIndexAsync(cancellationToken).ConfigureAwait(false);
```

Comment:
this worries me.  on a large solution where this hasn't been computed, won't this be super expensive?  i think we should only use the index if we've computed the index.

Reply:
Yes, that's the item2 I mentioned above, i.e.

> Determine the bail condition (e.g. if any of the filter items isn't calculated? And/Or a timebox? )

So editor team has implemented a "gold bar" for completion list, which shows something like "intellisense is not fully loaded" in certain scenarios, I think we might be able to use that

Reply:
OK, now we bail if we don't have all the filters ready (except for documents in current project, we still force build the filter for them). 

Also found an issue in SyntaxTreeIndex cache, fixed here https://github.com/dotnet/roslyn/pull/39126/commits/67263d696b98da1f79bd19be61332f22e5940cc9